### PR TITLE
GH-36099: [C++] Add Utf8View and BinaryView to the c ABI

### LINKED ISSUE #36099: Add string view to the C ABI
### Describe the enhancement requested

After https://github.com/apache/arrow/issues/35627 it will be useful to add the new types to the C ABI.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by bkietz**
### Rationale for this change

Utf8View and BinaryView should be added to the c ABI spec and to the c++ library's importer/exporter.

### Are these changes tested?

Yes, minimally

### Are there any user-facing changes?

View arrays will be importable/exportable through the c ABI in c++

* Closes: #36099

### CODE REVIEW

File: cpp/src/arrow/c/bridge_test.cc
Comment (bkietz) [inline cpp/src/arrow/c/bridge_test.cc:358]: Should I add a test exercising many/no buffers?

**Review by bkietz — COMMENTED** [review]
Note: based on https://github.com/apache/arrow/pull/37792

**Comment by felipecrv** [comment]
LGTM. But of course we need to wait for the separate review and merge of the string view PR before merging this.

File: cpp/src/arrow/c/bridge_test.cc
Comment (pitrou) [reply cpp/src/arrow/c/bridge_test.cc:358]: That would be nicer IMHO.

**Comment by pitrou** [comment]
You also need to remove the corresponding skips in datagen.py.

**Comment by bkietz** [comment]
The largest issue I had forgotten: C-ABI doesn't store buffer sizes explicitly. (For string data we derive the data buffer's length by reading the offsets buffer.) I guess we'll need an extra buffer to store those, since reconstructing the sizes by examining every view is untenable

**Comment by bkietz** [comment]
ping @pitrou

File: cpp/src/arrow/c/bridge.cc
Comment (zeroshade) [inline cpp/src/arrow/c/bridge.cc:267]: i've actually wondered.... is there any particular difference (performance or otherwise) doing `.reset(new T)` vs `= std::make_unique<T>()`?

File: cpp/src/arrow/c/bridge.cc
Comment (zeroshade) [inline cpp/src/arrow/c/bridge.cc:596]: the `size` here is a reference? is there any way for it to be more explicit that this is a reference in the syntax here?

File: cpp/src/arrow/c/bridge.cc
Comment (zeroshade) [inline cpp/src/arrow/c/bridge.cc:1423]: we lost the default to `DeviceAllocationType::kCPU`, should it be added to the initialization below?

File: cpp/src/arrow/c/bridge_test.cc
Comment (zeroshade) [inline cpp/src/arrow/c/bridge_test.cc:128]: should this be `{ false };` ? (unless I'm forgetting how the default initialization works)

**Review by zeroshade — COMMENTED** [review]
just a few nitpicks / comments

File: cpp/src/arrow/c/bridge.cc
Comment (felipecrv) [reply cpp/src/arrow/c/bridge.cc:267]: In scenarios where you have an `std::unique_ptr<ParentClass>` being populated by many sub-classes of `ParentClass`:

`std::make_unique<T, Deleter>(...)` can lead to more binary bloat because `class Deleter = std::default_delete<T>` is type-specialized whereas `reset()` takes a super-class pointer and only needs the deleter that invokes the destructor via dynamic dispatching through the parent class vtable.

Being type-specialized often leads to the derived-classes destructors' being inlined into the `std::default_deleter<T>`. All this extra inlining and removal of one indirection could lead to less overhead of destructor dispatching.

File: cpp/src/arrow/c/bridge.cc
Comment (pitrou) [reply cpp/src/arrow/c/bridge.cc:596]: Yes, or perhaps simply avoid `Zip` here and use the more usual and more readable push-back approach:
```c++
      export_.variadic_buffer_sizes_.reserve(variadic_buffers.size());
      for (const auto& buf : variadic_buffers) {
        export_.variadic_buffer_sizes_.push_back(buf->size());
      }

File: cpp/src/arrow/c/bridge_test.cc
Comment (pitrou) [reply cpp/src/arrow/c/bridge_test.cc:128]: It would be more readable like that, yes.

File: cpp/src/arrow/c/bridge_test.cc
Comment (pitrou) [inline cpp/src/arrow/c/bridge_test.cc:292]: Readability
```suggestion
  size_t flattened_index_{0};
```

File: cpp/src/arrow/c/bridge_test.cc
Comment (pitrou) [inline cpp/src/arrow/c/bridge_test.cc:930]: Why not use `TestPrimitive`? Also, ideally the factory creates the array on the fly, to make sure that lifetimes are handled properly.

**Review by pitrou — CHANGES_REQUESTED** [review]
1. I see no array roundtrip test, am I missing something?
2. Can you rebase to get the latest changes (including list view integration)?

File: cpp/src/arrow/c/bridge.cc
Comment (bkietz) [reply cpp/src/arrow/c/bridge.cc:1423]: This was a linter action, which replaced DeviceAllocationType::kCPU with `{}` below since that's equivalent in this case. I'll replace `{}` with the more explicit form.

Comment (pitrou) [reply cpp/src/arrow/c/bridge.cc:1423]: That's an aggressive linter...

File: cpp/src/arrow/c/bridge_test.cc
Comment (bkietz) [reply cpp/src/arrow/c/bridge_test.cc:930]: I'll use TestPrimitive and copy those buffers onto the heap

File: cpp/src/arrow/c/bridge.cc
Comment (zeroshade) [reply cpp/src/arrow/c/bridge.cc:267]: In that case, wouldn't it be better to keep this as calling `.reset` rather than switch to `make_unique`?

File: cpp/src/arrow/c/bridge.cc
Comment (zeroshade) [reply cpp/src/arrow/c/bridge.cc:1423]: i agree, that really is an aggressive linter.... lol

**Comment by bkietz** [comment]
@pitrou rebased, added round trip tests

**Review by pitrou — APPROVED** [review]
LGTM, but I think there are lint errors to fix

**Comment by felipecrv** [comment]
Yay! 🚀

**Comment by urvishdesai** [comment]
Hi @bkietz, is there a plan to add Utf8View support to the rust library soon? 

If someone is not already working on it, I would like to take up the PR to add support for Utf8View to the arrow2 library (https://github.com/jorgecarleitao/arrow2/issues/1596)

**Comment by bkietz** [comment]
@urvishdesai thanks for offering! I don't think anyone else has that queued, please do (and I'd be happy to review)

**Pull request merged**
